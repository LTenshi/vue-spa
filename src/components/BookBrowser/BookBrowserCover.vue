<template>
  <tr>
    <th class="imageHeader">
        <img :src="bookCover" @click="openDetails" class="imageContainer">
    </th>
    <th class="bookName">
        <div>{{bookName}}</div>
        <div class="bookAuthor">{{bookAuthor}}</div>
    </th>
  </tr>
</template>

<script>
export default {
    props: {
        bookName: {
            type: String,
            required: true,
        },
        bookAuthor: {
            type: String,
            required: true,
        },
        bookCover: {
            type: String,
            required: true,
        },
    },
    methods: {
        openDetails() {
            this.$emit("openDetails");
        }
    }
}
</script>

<style>
.bookName, .bookAuthor {
    font-size: 13pt;
}

.bookName {
    vertical-align: top;
    padding: 20px;
    text-align: left;
}

.bookAuthor {
    font-weight: 400;
    color: rgb(175, 175, 175);
}

.imageContainer {
    max-width: 140px;
}
</style>