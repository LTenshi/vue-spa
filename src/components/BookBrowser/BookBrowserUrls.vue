<template>
  <th>
    <a v-for="link in bookLinks" :key="link.name" :href="link.url" class="bookLink">{{link.name}}<br></a>
  </th>
</template>

<script>
export default {
  props: {
    bookLinks: {
      type: Array,
      default: () => []
    }
  }
}
</script>

<style>
.bookLink {
  color: rgb(136, 221, 0);
  text-decoration: none;
}

.bookLink > a:link {
  text-decoration: none;
}

.bookLink > a:visited {
  text-decoration: none;
}

.bookLink > a:hover {
  text-decoration: none;
}

.bookLink > a:active {
  text-decoration: none;
}
</style>