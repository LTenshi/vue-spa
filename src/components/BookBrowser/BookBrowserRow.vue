<template>
  <tr>
    <book-browser-cover 
        :bookCover="bookCover" 
        :bookName="bookName" 
        :bookAuthor="bookAuthor"
        @openDetails="openDetails"/>
    <th class="fontStyle">
        {{this.bookYear}}
    </th>
    <th class="fontStyle">
        {{this.bookRating}}/10
    </th>
    <book-browser-urls :bookLinks="bookLinks"/>
  </tr>
  <th v-if="detailsExpanded" colspan="4" class="detailsPanel" v-html="this.bookDetails"></th>
</template>

<script>
import BookBrowserCover from './BookBrowserCover.vue';
import BookBrowserUrls from './BookBrowserUrls.vue';
import showdown from "showdown";
export default {
    components: { BookBrowserCover, BookBrowserUrls },
    data() {
        return {
            detailsExpanded: false
        }
    },
    props: {
        book: {
            type: Object,
            required: true,
        },
    },
    computed: {
        bookCover() {
            return this.book.coverUrl;
        },
        bookName() {
            return this.book.name;
        },
        bookId() {
            return this.book.id;
        },
        bookAuthor() {
            return this.book.author;
        },
        bookYear() {
            return this.book.year;
        },
        bookRating() {
            return this.book.rating;
        },
        bookLinks() {
            return this.book.links;
        },
        bookDetails() {
            var converter = new showdown.Converter();
            return converter.makeHtml(this.book.details);
        }
    },
    methods: {
        openDetails() {
            this.detailsExpanded = !this.detailsExpanded;
        }
    }
}
</script>

<style>
.detailsPanel {
    font-weight: 450;
    text-align: initial;
}

.fontStyle {
    font-weight: 450;
}
</style>