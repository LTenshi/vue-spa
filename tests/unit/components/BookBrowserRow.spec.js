import BookBrowserRow from "../../../src/components/BookBrowser/BookBrowserRow";
import BookBrowserCover from "../../../src/components/BookBrowser/BookBrowserCover";
import BookBrowserUrls from "../../../src/components/BookBrowser/BookBrowserUrls";

import { mount } from "@vue/test-utils";

describe("Book Browser Cover", () => {
    let wrapper;
    let coverElement;
    let urlElement;

    beforeEach(() => {
        wrapper = mount(BookBrowserRow, {
            props: {
               book:  
                {
                    "id": "0",
                    "coverUrl": "https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fcdn2.penguin.com.au%2Fcovers%2Foriginal%2F9780141956794.jpg&f=1&nofb=1",
                    "name": "In Search of Lost Time",
                    "author": "Marcel Proust",
                    "year": "1913",
                    "rating": "9.8",
                    "links": [
                    { "name": "Amazon", "url": "https://www.amazon.co.uk" },
                    { "name": "iBooks", "url": "https://www.apple.com/uk/apple-books/" },
                    { "name": "Play Store", "url": "https://play.google.com/store/books" }
                    ],
                    "details": "**In Search of Lost Time** (French: Ã€ la recherche du temps perdu), first translated into English as Remembrance of Things Past, and sometimes referred to in French as La Recherche (The Search), is a novel in seven volumes by French author Marcel Proust.This early 20th-century work is his most prominent, known both for its length and its theme of involuntary memory."
                },
            }
        });
        coverElement = wrapper.findComponent(BookBrowserCover);
        urlElement = wrapper.findComponent(BookBrowserUrls);
    })

    it("Should match the snapshot", async () => {
        expect(wrapper.element).toMatchSnapshot();

        await coverElement.vm.openDetails();

        expect(wrapper.element).toMatchSnapshot("BookBrowserRowExpanded");
    })
    
    it('Should contain a cover element', () => {
        expect(coverElement.exists()).toBe(true);
    });
    
    it('Should contain a url element', () => {
        expect(urlElement.exists()).toBe(true);
    });

    it('Should contain a book year', () => {
        expect(wrapper.html()).toContain("1913");
    });

    it('Should contain a score', () => {
        expect(wrapper.html()).toContain("9.8/10");
    });

    it('Should not display details', () => {
        expect(wrapper.html()).not.toContain("**In Search of Lost Time**");
    });

    it('Should display details when image is clicked', async () => {
        await coverElement.vm.openDetails();

        expect(wrapper.html()).toContain("<strong>In Search of Lost Time</strong>");
    });
    
});