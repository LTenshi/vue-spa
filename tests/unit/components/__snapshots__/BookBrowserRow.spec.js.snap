// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book Browser Cover Should match the snapshot 1`] = `
<div
  data-v-app=""
>
  
  <tr>
    <tr>
      <th
        class="imageHeader"
      >
        <img
          class="imageContainer"
          src="https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fcdn2.penguin.com.au%2Fcovers%2Foriginal%2F9780141956794.jpg&f=1&nofb=1"
        />
      </th>
      <th
        class="bookName"
      >
        <div>
          In Search of Lost Time
        </div>
        <div
          class="bookAuthor"
        >
          Marcel Proust
        </div>
      </th>
    </tr>
    <th
      class="fontStyle"
    >
      1913
    </th>
    <th
      class="fontStyle"
    >
      9.8/10 
    </th>
    <th>
      
      <a
        class="bookLink"
        href="https://www.amazon.co.uk"
      >
        Amazon
        <br />
      </a>
      <a
        class="bookLink"
        href="https://www.apple.com/uk/apple-books/"
      >
        iBooks
        <br />
      </a>
      <a
        class="bookLink"
        href="https://play.google.com/store/books"
      >
        Play Store
        <br />
      </a>
      
    </th>
  </tr>
  <!--v-if-->
  
</div>
`;

exports[`Book Browser Cover Should match the snapshot: BookBrowserRowExpanded 1`] = `
<div
  data-v-app=""
>
  
  <tr>
    <tr>
      <th
        class="imageHeader"
      >
        <img
          class="imageContainer"
          src="https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fcdn2.penguin.com.au%2Fcovers%2Foriginal%2F9780141956794.jpg&f=1&nofb=1"
        />
      </th>
      <th
        class="bookName"
      >
        <div>
          In Search of Lost Time
        </div>
        <div
          class="bookAuthor"
        >
          Marcel Proust
        </div>
      </th>
    </tr>
    <th
      class="fontStyle"
    >
      1913
    </th>
    <th
      class="fontStyle"
    >
      9.8/10 
    </th>
    <th>
      
      <a
        class="bookLink"
        href="https://www.amazon.co.uk"
      >
        Amazon
        <br />
      </a>
      <a
        class="bookLink"
        href="https://www.apple.com/uk/apple-books/"
      >
        iBooks
        <br />
      </a>
      <a
        class="bookLink"
        href="https://play.google.com/store/books"
      >
        Play Store
        <br />
      </a>
      
    </th>
  </tr>
  <th
    class="detailsPanel"
    colspan="4"
  >
    <p>
      <strong>
        In Search of Lost Time
      </strong>
       (French: Ã€ la recherche du temps perdu), first translated into English as Remembrance of Things Past, and sometimes referred to in French as La Recherche (The Search), is a novel in seven volumes by French author Marcel Proust.This early 20th-century work is his most prominent, known both for its length and its theme of involuntary memory.
    </p>
  </th>
  
</div>
`;
